### ISSUE TITLE
Review Lab n.10 - Tic Tac Toe RL - by Ivan Magistro Contenta s314356

### ISSUE THREAD

**Issue by ivanmag22** [arturoadelfio/computational_intelligence#5]
Hi Arturo!
I've seen your work and, even if it isn't complete, your implementation and idea of using Q-learning are good.
I want to do some considerations and to tell you some suggestions:

- it is useful to **print** not only the *wins* and the *losses*, but also the *draws*. I suggest to put ```result = state_value(state)``` outside the while loop to get the result of the game.
With your approach I obtained the following results (they aren't too bad, but we can do better!):
  - *won games*: 0.54 around
  - *draw games*: 0.17 around
  - *lost games*: 0.29 around
- **plots** are also useful because they are intuitive
- as regard the **hyper-parameters** I suggest to use a higher value for *learning rate*, otherwise you proceed with very small steps to the best combination of moves to reach the victory. Moreover I suggest to try different values of *exploration probability* and to implement better that part in order to pick the action according to a policy (I understood that you wanted to use a epsilon-policy that is good) and not randomly.
- I advice to you to handle well the *Q-table*: it could bring you to select the best actions according to Q-table values.
- I appreciated the way you managed all with *class and methods* for both QAgent and RandomPlayer.
- I appreciated the fact that you give **rewards** not only for *wins* or *losses*, but also to the *most promising moves* (with magic number square the most promising are the highest ones in order to reach in a short number of moves the sum of 15).

Good work!
Good luck for the exam :)
